### PR TITLE
Improving the 64-bit number formatting to better match the 32-bit algorithm

### PR DESCRIPTION
**Pull request by tannergooding**
Found that the implementation was less efficient than expected when working on `Int128` and `UInt128`. This updates the 64-bit paths to be closer to what the 32-bit paths do, but with keeping the "decomposed" operation on 32-bit platforms.

### CODE REVIEW

**Comment by danmoseley** [comment]
Do you plan to add a few int128 formatting/parsing benchmarks in the perf repo?

**Comment by tannergooding** [comment]
> Do you plan to add a few int128 formatting/parsing benchmarks in the perf repo?

After Int128/UInt128 is added to the repo, yes. This is just some intermediate cleanup that will make adding those types simpler.

I'll get benchmark results after CI is passing.

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs [hunk 1]
```diff
@@ -1171,7 +1171,7 @@ static unsafe bool TryFormatUInt64Slow(ulong value, ReadOnlySpan<char> format, I
             }
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)] // called from only one location
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:1174]: Was the comment stale, or this PR is causing it to be called in more places?

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs [hunk 2]
```diff
@@ -1297,6 +1297,7 @@ private static unsafe bool TryInt32ToHexStr(int value, char hexBase, int digits,
             return true;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:1300]: We're ok with whatever additional code size this leads to? (General question for the AggressiveInlinings added in the PR)

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs [hunk 3]
```diff
@@ -1608,12 +1623,112 @@ private static unsafe void UInt64ToNumber(ulong value, ref NumberBuffer number)
...
+            while ((uint)(value >> 32) != 0)
+            {
+                bufferEnd = UInt32ToDecChars(bufferEnd, Int64DivMod1E9(ref value), 9);
+            }
+            return UInt32ToDecChars(bufferEnd, (uint)value);
+#else
+            do
+            {
+                ulong remainder;
+                (value, remainder) = Math.DivRem(value, 10);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:1649]: We use Math.DivRem here and elsewhere but separate % and / operations above in Int64DivMod1E9, and in both cases the divisor is a constant.  Is there a reason for the difference?

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs [hunk 4]
```diff
@@ -1608,12 +1623,112 @@ private static unsafe void UInt64ToNumber(ulong value, ref NumberBuffer number)
...
+                *(--bufferEnd) = (char)(remainder + '0');
+            }
+            return bufferEnd;
+#endif
+        }
+
+        internal static unsafe string UInt64ToDecStr(ulong value)
+        {
+            // Intrinsified in mono interpreter
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:1730]: What does this comment refer to?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:1174]: Yes. We now handle both `byte` and `char` based versions so its at least 2 places

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs (hunk 3 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:1649]: `DivRem` is more efficient. `Int64DivMod1E9` is a bit of a special case where we are effectively doing both `32-bit = 64-bit % 32-bit` and `64-bit = 64-bit / 32-bit`. 

The former can be optimized on some platforms, but I didn't do any investigation to confirm if the JIT was doing anything special here or not.

File: src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs (hunk 4 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:1730]: I believe its referring to `CountDigits`. It's a pre-existing comment that I preserved

**Review by dakersnar — APPROVED** [review]
LGTM

**Comment by AndyAyersMS** [comment]
Improvements seen on arm64: https://github.com/dotnet/perf-autofiling-issues/issues/5277

**Pull request merged**